File: .opencode/tool/validate_notebook.py
#!/usr/bin/env python3
"""Notebook Validation Tool for ADW.

Validates Jupyter notebook structure and syntax without executing code and
adds Jupytext-powered conversion and sync operations. Supports validation for
single files or directories (optional recursion) plus conversion (`--convert-to-py`),
bidirectional sync (`--sync`), and CI read-only sync checks (`--check-sync`).

Exit codes:
  - 0: success (valid notebooks / conversions / sync / in-sync)
  - 1: functional failure (invalid notebooks, conversion failures, out-of-sync)
  - 2: tool/argument error

Examples:
  # Validation
  python3 .opencode/tool/validate_notebook.py docs/Examples --recursive --output json
  # Conversion
  python3 .opencode/tool/validate_notebook.py notebook.ipynb --convert-to-py
  python3 .opencode/tool/validate_notebook.py docs/Examples --recursive --convert-to-py --output-dir scripts
  # Sync
  python3 .opencode/tool/validate_notebook.py notebook.ipynb --sync
  # Check-sync (read-only)
  python3 .opencode/tool/validate_notebook.py docs/Examples --recursive --check-sync
"""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path
from typing import TYPE_CHECKING, Iterable, List, Optional, Sequence

REPO_ROOT = Path(__file__).resolve().parents[2]
if str(REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(REPO_ROOT))

# Lazy imports to avoid unnecessary dependencies for conversion/sync modes
if TYPE_CHECKING:
    from adw.utils.notebook_validation import NotebookValidationResult

MAX_ERRORS_PER_NOTEBOOK = 5
FULL_OUTPUT_CHAR_LIMIT = 4000
CHECK_SYNC_MTIME_TOLERANCE = 0.0


class ValidationToolError(Exception):
    """Predictable error for user-facing failures."""


def _is_hidden(path: Path) -> bool:
    """Check whether any segment of the path is hidden.

    Args:
        path: Filesystem path to inspect.

    Returns:
        True when any segment starts with a dot, otherwise False.
    """

    return any(part.startswith(".") for part in path.parts)


def _collect_notebooks(target: Path, recursive: bool) -> List[Path]:
    """Collect notebook paths from a file or directory.

    Skips hidden paths and .ipynb_checkpoints, returns deterministically sorted list.
    Hidden paths are defined by any segment starting with a dot.
    """

    if target.is_file() and target.suffix == ".ipynb":
        return [target]

    if not target.exists():
        raise ValidationToolError(f"Path not found: {target}")

    if not target.is_dir():
        raise ValidationToolError(f"Path must be .ipynb file or directory: {target}")

    candidates: Iterable[Path]
    if recursive:
        candidates = target.rglob("*.ipynb")
    else:
        candidates = target.glob("*.ipynb")

    notebooks = [p for p in candidates if ".ipynb_checkpoints" not in p.parts and not _is_hidden(p)]

    if not notebooks:
        raise ValidationToolError(f"No notebooks found under {target}")

    return sorted(notebooks)


def _truncate(text: str, limit: int = FULL_OUTPUT_CHAR_LIMIT) -> tuple[str, bool]:
    """Truncate text to the limit, returning text and truncation flag.

    Args:
        text: Content to truncate.
        limit: Maximum allowed characters; defaults to FULL_OUTPUT_CHAR_LIMIT.

    Returns:
        Tuple of truncated text and a boolean indicating whether truncation occurred.
    """

    if len(text) <= limit:
        return text, False
    return f"{text[:limit]}\n...\n[output truncated to {limit} chars]", True


def _format_summary(results: List[NotebookValidationResult]) -> str:
    """Return summary text for a batch of notebook validation results.

    Args:
        results: List of notebook validation results.

    Returns:
        Human-readable summary string with counts and failures.
    """

    checked = len(results)
    valid = sum(1 for r in results if r.valid)
    invalid = checked - valid

    lines: List[str] = []
    lines.append("=" * 60)
    lines.append("NOTEBOOK VALIDATION SUMMARY")
    lines.append("=" * 60)
    lines.append(f"\nNotebooks Checked: {checked}")
    lines.append(f"  Valid:   {valid}")
    if invalid:
        lines.append(f"  Invalid: {invalid}")
        lines.append(f"\nFailures ({invalid}):")
        for result in results:
            if result.valid:
                continue
            lines.append(f"  - {Path(result.notebook_path).name}")
            for error in result.errors[:MAX_ERRORS_PER_NOTEBOOK]:
                lines.append(f"      {error.error_type}: {error.message}")
            remaining = max(len(result.errors) - MAX_ERRORS_PER_NOTEBOOK, 0)
            if remaining:
                lines.append(f"      ... and {remaining} more errors")

    lines.append("\n" + "=" * 60)
    lines.append("VALIDATION: PASSED" if invalid == 0 else "VALIDATION: FAILED")
    lines.append("=" * 60)
    return "\n".join(lines)


def _format_full(results: List[NotebookValidationResult], truncated_notice: bool = True) -> str:
    """Return detailed text for each notebook, optionally noting truncation.

    Args:
        results: List of notebook validation results.
        truncated_notice: Whether to append a note when output is truncated.

    Returns:
        String containing detailed notebook validation information.
    """

    lines: List[str] = []
    for result in results:
        lines.append("=" * 60)
        lines.append(f"Notebook: {result.notebook_path}")
        lines.append("=" * 60)
        lines.append(f"Valid: {result.valid}")
        if result.warnings:
            lines.append("Warnings:")
            for warning in result.warnings:
                lines.append(f"  - {warning}")
        if result.errors:
            lines.append("Errors:")
            for error in result.errors:
                lines.append(f"  - {error.error_type}: {error.message}")
                if error.cell_index is not None:
                    lines.append(f"      cell_index: {error.cell_index}")
                if error.details:
                    for key, value in error.details.items():
                        lines.append(f"      {key}: {value}")
        lines.append("")

    combined = "\n".join(lines)
    truncated_text, truncated = _truncate(combined)
    if truncated and truncated_notice:
        return f"{truncated_text}\n[output was truncated]"
    return truncated_text


def _format_json(results: List[NotebookValidationResult], skip_syntax: bool) -> str:
    """Return JSON string with validation results and meta fields.

    Args:
        results: List of notebook validation results.
        skip_syntax: Whether syntax errors were converted to warnings.

    Returns:
        JSON-formatted string describing validation outcomes.
    """

    payload = {
        "notebooks_checked": len(results),
        "notebooks_valid": sum(1 for r in results if r.valid),
        "notebooks_invalid": sum(1 for r in results if not r.valid),
        "skip_syntax": skip_syntax,
        "results": [],
    }

    serialized = []
    for result in results:
        serialized_errors = [
            {
                "error_type": err.error_type,
                "message": err.message,
                "cell_index": err.cell_index,
                "details": err.details,
            }
            for err in result.errors
        ]
        serialized.append(
            {
                "notebook_path": result.notebook_path,
                "valid": result.valid,
                "errors": serialized_errors,
                "warnings": result.warnings,
            }
        )
    payload["results"] = serialized
    return json.dumps(payload, indent=2)


def _validate_output_dir_usage(args: argparse.Namespace) -> None:
    if args.output_dir and not args.convert_to_py:
        raise ValidationToolError("--output-dir is only allowed with --convert-to-py")


def _ensure_validation_only_flags(args: argparse.Namespace) -> None:
    if (args.output != "summary" or args.skip_syntax) and (
        args.convert_to_py or args.sync or args.check_sync
    ):
        raise ValidationToolError(
            "--output/--skip-syntax can only be used with validation (omit convert/sync flags)"
        )


def _handle_convert(args: argparse.Namespace) -> int:
    from adw.utils.notebook_jupytext import SyncResult, notebook_to_script

    target = Path(args.notebook_path)
    notebooks = _collect_notebooks(target, recursive=args.recursive)
    output_dir = Path(args.output_dir) if args.output_dir else None

    failures: list[SyncResult] = []
    for notebook in notebooks:
        output_path = None
        if output_dir:
            output_path = output_dir / notebook.with_suffix(".py").name
        result = notebook_to_script(notebook, output_path)
        status = "✓" if result.success else "✗"
        action = result.action
        message = f"{status} {notebook} -> {result.target_path} ({action})"
        if result.error_message:
            message = f"{message}: {result.error_message}"
        print(message)
        if not result.success:
            failures.append(result)

    if failures:
        return 1
    return 0


def _handle_sync(args: argparse.Namespace) -> int:
    from adw.utils.notebook_jupytext import SyncResult, sync_notebook_script

    if args.output_dir:
        raise ValidationToolError("--output-dir is only allowed with --convert-to-py")

    target = Path(args.notebook_path)
    notebooks = _collect_notebooks(target, recursive=args.recursive)

    failures: list[SyncResult] = []
    for notebook in notebooks:
        result = sync_notebook_script(notebook)
        status = "✓" if result.success else "✗"
        message = f"{status} {notebook} -> {result.target_path} ({result.action})"
        if result.error_message:
            message = f"{message}: {result.error_message}"
        print(message)
        if not result.success:
            failures.append(result)

    if failures:
        return 1
    return 0


def _handle_check_sync(args: argparse.Namespace) -> int:
    if args.output_dir:
        raise ValidationToolError("--output-dir is only allowed with --convert-to-py")

    target = Path(args.notebook_path)
    notebooks = _collect_notebooks(target, recursive=args.recursive)

    out_of_sync: list[str] = []
    for notebook in notebooks:
        script_path = notebook.with_suffix(".py")
        if not script_path.exists():
            out_of_sync.append(f"{notebook}: missing script")
            continue
        try:
            nb_mtime = notebook.stat().st_mtime_ns
            py_mtime = script_path.stat().st_mtime_ns
        except FileNotFoundError:
            out_of_sync.append(f"{notebook}: failed to stat files")
            continue

        mtime_diff = nb_mtime - py_mtime
        if abs(mtime_diff) <= CHECK_SYNC_MTIME_TOLERANCE:
            continue
        if mtime_diff > 0:
            out_of_sync.append(f"{notebook}: notebook is newer")
        else:
            out_of_sync.append(f"{notebook}: script is newer")

    if out_of_sync:
        print("Out of sync:")
        for line in out_of_sync:
            print(f"  ✗ {line}")
        return 1

    print(f"All {len(notebooks)} notebooks are in sync with their scripts")
    return 0


def main(argv: Optional[Sequence[str]] = None) -> int:
    """CLI entrypoint with validation, conversion, sync, and check-sync modes."""

    parser = argparse.ArgumentParser(
        description=("Validate notebooks or perform Jupytext conversion/sync operations."),
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=(
            "Examples:\n"
            "  # Validation\n"
            "  python3 .opencode/tool/validate_notebook.py notebook.ipynb\n"
            "  python3 .opencode/tool/validate_notebook.py docs/Examples --recursive --output json\n"
            "\n"
            "  # Conversion\n"
            "  python3 .opencode/tool/validate_notebook.py notebook.ipynb --convert-to-py\n"
            "  python3 .opencode/tool/validate_notebook.py docs/Examples --recursive --convert-to-py --output-dir scripts\n"
            "\n"
            "  # Sync\n"
            "  python3 .opencode/tool/validate_notebook.py notebook.ipynb --sync\n"
            "  python3 .opencode/tool/validate_notebook.py docs/Examples --recursive --check-sync\n"
        ),
    )

    parser.add_argument("notebook_path", help="Path to notebook file or directory")
    parser.add_argument(
        "--recursive",
        action="store_true",
        help="Search directories recursively for .ipynb files",
    )

    parser.add_argument(
        "--output",
        choices=["summary", "full", "json"],
        default="summary",
        help="Output mode (validation only; default: summary)",
    )
    parser.add_argument(
        "--skip-syntax",
        action="store_true",
        help=("Skip Python syntax checking (validation only); syntax issues become warnings"),
    )

    action_group = parser.add_mutually_exclusive_group()
    action_group.add_argument(
        "--convert-to-py",
        dest="convert_to_py",
        action="store_true",
        help="Convert notebooks to .py:percent format",
    )
    action_group.add_argument(
        "--sync",
        action="store_true",
        help="Bidirectional sync between notebook and script (newer file wins)",
    )
    action_group.add_argument(
        "--check-sync",
        dest="check_sync",
        action="store_true",
        help="Check notebook/script sync state (read-only, CI friendly)",
    )

    parser.add_argument(
        "--output-dir",
        type=str,
        help="Output directory for converted files (only with --convert-to-py)",
    )

    try:
        args = parser.parse_args(list(argv) if argv is not None else None)
    except SystemExit as exc:
        code = exc.code
        return code if isinstance(code, int) else 2

    try:
        _validate_output_dir_usage(args)
        _ensure_validation_only_flags(args)

        if args.convert_to_py:
            return _handle_convert(args)
        if args.sync:
            return _handle_sync(args)
        if args.check_sync:
            return _handle_check_sync(args)

        # Validation path
        from adw.utils.notebook_validation import validate_notebook_json

        target = Path(args.notebook_path)
        notebooks = _collect_notebooks(target, recursive=args.recursive)
        results = [validate_notebook_json(nb, skip_syntax=args.skip_syntax) for nb in notebooks]

        if args.output == "json":
            output = _format_json(results, skip_syntax=args.skip_syntax)
        elif args.output == "full":
            output = _format_full(results)
        else:
            output = _format_summary(results)

        print(output)
        all_valid = all(r.valid for r in results)
        return 0 if all_valid else 1
    except ValidationToolError as exc:
        print(f"ERROR: {exc}")
        return 2
    except Exception as exc:  # noqa: BLE001
        print(f"ERROR: Unexpected failure: {exc}")
        return 2


if __name__ == "__main__":
    sys.exit(main())

File: .opencode/tools/validate_notebook.ts
/**
 * Notebook Validation Tool
 *
 * Validates Jupyter notebook structure and syntax without execution.
 * Mirrors run_notebook patterns with summary/full/json modes.
 *
 * Exit codes: 0=all valid, 1=invalid notebooks, 2=tool error.
 */

import { tool } from "@opencode-ai/plugin";

const MISSING_SCRIPT_HINT = "Missing backing script .opencode/tools/validate_notebook.py";

type OutputMode = "summary" | "full" | "json";
type ValidationMode = "fast" | "full";

type ParsedValidationOptions = {
  outputMode?: OutputMode;
  skipSyntax?: true;
  validationMode?: ValidationMode;
  fast?: true;
  full?: true;
  explicitTokens?: true;
};

type ParsedValidationOptionsResult =
  | { ok: true; options: ParsedValidationOptions }
  | { ok: false; error: string };

const VALIDATION_OPTION_RULES = new Set([
  "output",
  "skip-syntax",
  "validation-mode",
  "fast",
  "full",
]);

const OUTPUT_MODES = new Set<OutputMode>(["summary", "full", "json"]);
const VALIDATION_MODES = new Set<ValidationMode>(["fast", "full"]);

const tokenizeOptions = (options: string): { ok: true; tokens: string[] } | { ok: false; error: string } => {
  const tokens: string[] = [];
  let current = "";
  let quote: "'" | '"' | undefined;

  for (let index = 0; index < options.length; index += 1) {
    const char = options[index];
    if (quote) {
      current += char;
      if (char === quote) {
        quote = undefined;
      }
      continue;
    }

    if (char === "'" || char === '"') {
      quote = char;
      current += char;
      continue;
    }

    if (/\s/.test(char)) {
      if (current) {
        tokens.push(current);
        current = "";
      }
      continue;
    }

    current += char;
  }

  if (quote) {
    return { ok: false, error: "ERROR: Invalid options string: unterminated quoted value." };
  }
  if (current) {
    tokens.push(current);
  }

  return { ok: true, tokens };
};

const stripOptionalQuotes = (value: string): string => {
  if (value.length >= 2) {
    const first = value[0];
    const last = value[value.length - 1];
    if ((first === '"' || first === "'") && last === first) {
      return value.slice(1, -1);
    }
  }

  return value;
};

const parseValidationOptions = (options: unknown): ParsedValidationOptionsResult => {
  if (options === undefined || options === null) {
    return { ok: true, options: {} };
  }
  if (typeof options !== "string") {
    return { ok: false, error: "ERROR: 'options' must be a string when provided." };
  }

  const normalized = options.trim();
  if (!normalized) {
    return { ok: true, options: {} };
  }

  const tokenized = tokenizeOptions(normalized);
  if (!tokenized.ok) {
    return tokenized;
  }

  const parsed: ParsedValidationOptions = { explicitTokens: true };
  for (const token of tokenized.tokens) {
    const separatorIndex = token.indexOf("=");
    if (separatorIndex !== token.lastIndexOf("=")) {
      return {
        ok: false,
        error: `ERROR: Invalid options token '${token}': tokens must contain at most one '=' separator.`,
      };
    }

    if (separatorIndex === -1) {
      if (!VALIDATION_OPTION_RULES.has(token)) {
        return { ok: false, error: `ERROR: Invalid options token '${token}': token is not supported.` };
      }
      if (token === "output" || token === "validation-mode") {
        return {
          ok: false,
          error: `ERROR: Invalid options token '${token}': token requires a non-empty '=value' suffix.`,
        };
      }
      if ((token === "skip-syntax" && parsed.skipSyntax)
        || (token === "fast" && parsed.fast)
        || (token === "full" && parsed.full)) {
        return { ok: false, error: `ERROR: Invalid options token '${token}': duplicate token.` };
      }
      if (token === "skip-syntax") {
        parsed.skipSyntax = true;
      } else if (token === "fast") {
        parsed.fast = true;
      } else if (token === "full") {
        parsed.full = true;
      }
      continue;
    }

    const name = token.slice(0, separatorIndex);
    const rawValue = token.slice(separatorIndex + 1);
    if (!VALIDATION_OPTION_RULES.has(name)) {
      return { ok: false, error: `ERROR: Invalid options token '${token}': token is not supported.` };
    }
    if (!rawValue) {
      return {
        ok: false,
        error: `ERROR: Invalid options token '${token}': token requires a non-empty '=value' suffix.`,
      };
    }
    if (name === "skip-syntax" || name === "fast" || name === "full") {
      return { ok: false, error: `ERROR: Invalid options token '${token}': token does not accept a value.` };
    }

    const value = stripOptionalQuotes(rawValue).trim();
    if (!value) {
      return {
        ok: false,
        error: `ERROR: Invalid options token '${token}': token requires a non-empty '=value' suffix.`,
      };
    }

    if (name === "output") {
      if (!OUTPUT_MODES.has(value as OutputMode)) {
        return {
          ok: false,
          error: `ERROR: Invalid options token '${token}': output must be one of summary, full, json.`,
        };
      }
      if (parsed.outputMode !== undefined) {
        return { ok: false, error: `ERROR: Invalid options token '${token}': duplicate token.` };
      }
      parsed.outputMode = value as OutputMode;
      continue;
    }

    if (!VALIDATION_MODES.has(value as ValidationMode)) {
      return {
        ok: false,
        error: `ERROR: Invalid options token '${token}': validation-mode must be one of fast, full.`,
      };
    }
    if (parsed.validationMode !== undefined) {
      return { ok: false, error: `ERROR: Invalid options token '${token}': duplicate token.` };
    }
    parsed.validationMode = value as ValidationMode;
  }

  return { ok: true, options: parsed };
};

export default tool({
  description: `Validate, convert, or sync Jupyter notebooks using Jupytext.

EXAMPLES:
- Validation: validate_notebook({notebookPath: 'notebook.ipynb'})
- Validation with options: validate_notebook({notebookPath: 'notebook.ipynb', options: 'output=json fast'})
- Convert: validate_notebook({notebookPath: 'notebook.ipynb', convertToPy: true})
- Convert with output dir: validate_notebook({notebookPath: 'docs/Examples', recursive: true, convertToPy: true, outputDir: 'scripts'})
- Convert to notebook: validate_notebook({notebookPath: 'script.py', convertToIpynb: true})
- Sync: validate_notebook({notebookPath: 'notebook.ipynb', sync: true})
- Check-sync (CI): validate_notebook({notebookPath: 'docs/Examples', recursive: true, checkSync: true})

Exit codes: 0=success, 1=functional failure (invalid/out-of-sync/convert failure), 2=tool error`,
  args: {
    notebookPath: tool.schema
      .string()
      .describe("Path to notebook file (.ipynb) or directory"),
    recursive: tool.schema
      .boolean()
      .optional()
      .describe("Search recursively when notebookPath is a directory"),
    options: tool.schema
      .string()
      .optional()
      .describe(
        "Validation-only bounded options: output=<summary|full|json>, skip-syntax, validation-mode=<fast|full>, fast, full",
      ),
    convertToPy: tool.schema
      .boolean()
      .optional()
      .describe("Convert notebooks to .py:percent format"),
    convertToIpynb: tool.schema
      .boolean()
      .optional()
      .describe("Convert py:percent scripts to .ipynb notebooks"),
    sync: tool.schema
      .boolean()
      .optional()
      .describe("Bidirectional sync between notebook and script (newer wins)"),
    checkSync: tool.schema
      .boolean()
      .optional()
      .describe("Check notebook/script sync state (read-only; exit 1 when out of sync)"),
    outputDir: tool.schema
      .string()
      .optional()
      .describe("Output directory for converted files (only with convertToPy or convertToIpynb)"),
  },
  async execute(args) {
    const notebookPath = typeof args.notebookPath === "string" ? args.notebookPath.trim() : "";
    if (!notebookPath) {
      return "ERROR: notebookPath is required and must be non-empty.";
    }

    const parsedOptions = parseValidationOptions(args.options);
    if (!parsedOptions.ok) {
      return parsedOptions.error;
    }

    const recursive = Boolean(args.recursive);
    const outputMode = parsedOptions.options.outputMode || "summary";
    const skipSyntax = parsedOptions.options.skipSyntax === true;
    const validationMode = parsedOptions.options.validationMode;
    const fast = parsedOptions.options.fast === true;
    const full = parsedOptions.options.full === true;
    const convertToPy = Boolean(args.convertToPy);
    const convertToIpynb = Boolean(args.convertToIpynb);
    const sync = Boolean(args.sync);
    const checkSync = Boolean(args.checkSync);
    const outputDir = typeof args.outputDir === "string" ? args.outputDir.trim() : undefined;

    if (validationMode && (fast || full)) {
      return "ERROR: Conflicting validation options: use either 'validation-mode' or 'fast/full' aliases, not both.";
    }
    if (fast && full) {
      return "ERROR: Conflicting validation options: 'fast' and 'full' cannot both be true.";
    }

    const isValidationAction = !convertToPy && !convertToIpynb && !sync && !checkSync;
    const usesValidationOnlyFlags =
      parsedOptions.options.explicitTokens === true
      || outputMode !== "summary"
      || skipSyntax
      || Boolean(validationMode)
      || fast
      || full;

    if (!isValidationAction && usesValidationOnlyFlags) {
      return (
        "ERROR: Validation-only bounded options (options: output=..., skip-syntax, validation-mode=..., fast, full) " +
        "can only be used with validation (omit convert/sync/checkSync flags)."
      );
    }

    const cmdParts: (string | number)[] = [
      "python3",
      `${import.meta.dir}/validate_notebook.py`,
      notebookPath,
    ];

    if (convertToPy) {
      cmdParts.push("--convert-to-py");
    }
    if (convertToIpynb) {
      cmdParts.push("--convert-to-ipynb");
    }
    if (sync) {
      cmdParts.push("--sync");
    }
    if (checkSync) {
      cmdParts.push("--check-sync");
    }
    if (outputDir) {
      cmdParts.push("--output-dir", outputDir);
    }

    if (isValidationAction) {
      cmdParts.push(`--output=${outputMode}`);
      if (skipSyntax) {
        cmdParts.push("--skip-syntax");
      }
      if (validationMode) {
        cmdParts.push("--validation-mode", validationMode);
      }
      if (fast) {
        cmdParts.push("--fast");
      }
      if (full) {
        cmdParts.push("--full");
      }
    }

    if (recursive) {
      cmdParts.push("--recursive");
    }

    try {
      const result = await Bun.$`${cmdParts}`.text();
      return result || "Notebook tool completed but returned no output.";
    } catch (error: any) {
      const stdout = error?.stdout?.toString?.() || "";
      const stderr = error?.stderr?.toString?.() || "";
      const message = error?.message || "Unknown error";
      const combinedLower = `${stderr} ${message}`.toLowerCase();

      if (stdout.trim()) {
        return stdout;
      }

      if (stderr.trim()) {
        const hint = combinedLower.includes("enoent") ? `\n${MISSING_SCRIPT_HINT}` : "";
        return `ERROR: Notebook tool failed\n\n${stderr}${hint}`;
      }

      if (combinedLower.includes("enoent")) {
        return `ERROR: Failed to run notebook tool: ${message}\n${MISSING_SCRIPT_HINT}`;
      }

      return `ERROR: Failed to run notebook tool: ${message}`;
    }
  },
});
